# Pre-Processing the 5 years of SEC data to add our custom target variable

##### Reusable Functions for later

In [ ]:
## unique items in dataframe
# unique_CIKs_count = df_MVP_16['CIK'].nunique()
# print(unique_CIKs_count)



# ### Dataset for 27 most usable columns
# list_of_cols_27 = ['FILING_DATE', 'SUBMISSION_TYPE', 'CIK', 'PROGRESSUPDATE', 'NATUREOFAMENDMENT','LEGALSTATUSFORM', 
#                    'JURISDICTIONORGANIZATION', 'DATEINCORPORATION', 'CITY', 'STATEORCOUNTRY', 
#                    'COMPANYNAME', 'SECURITYOFFEREDTYPE', 'OFFERINGAMOUNT',
#                    'MAXIMUMOFFERINGAMOUNT', 'CURRENTEMPLOYEES', 'TOTALASSETMOSTRECENTFISCALYEAR',
#                    'TOTALASSETPRIORFISCALYEAR', 'CASHEQUIMOSTRECENTFISCALYEAR', 'CASHEQUIPRIORFISCALYEAR',
#                    'ACTRECEIVEDRECENTFISCALYEAR', 'ACTRECEIVEDPRIORFISCALYEAR', 'SHORTTERMDEBTMRECENTFISCALYEAR',
#                    'SHORTTERMDEBTPRIORFISCALYEAR', 'LONGTERMDEBTRECENTFISCALYEAR', 'LONGTERMDEBTPRIORFISCALYEAR',
#                    'NETINCOMEMOSTRECENTFISCALYEAR', 'NETINCOMEPRIORFISCALYEAR', 'Funding_Success']

# df_28 = dfarget[list_of_cols_27]
# display(df_28.head(3))

### Save Excel to CSV
# path = "/root/Funders-USA-ML-Backend/datasets/datasets_SEC/Aggregated SEC Data CSV.csv"
# df_csv = df.to_csv(path)


### creating a list of rows used in the loop
# row_list = []
# for i in range(5):
#     row_list.append(row)
#     display(row_list)


### using map function on df
# df = df_MVP['Count_of_x'] = df[''].map(df['Color'].value_counts())
# display(test_df.tail(2))


### working example of using .iloc to index with dataframes
# display(df.iloc[0,1])
# display(df.iloc[2,18])
# df.iloc[2,18] = 1
# display(df.iloc[2,18])


### working example of using .append on a dataframe
# ndf = ndf.append(df.iloc[0,:])
# display(ndf.tail(5))


### working example of a custom count variable in a for loop
# count = 0
# for i in range(5):
#     count += 1
#     print(count)


### Dropping duplicates in dataframe
# df = df.drop_duplicates(subset = ['CIK', 'SUBMISSION_TYPE'] keep="first")
# path = "/root/Funders-USA-ML-Backend/datasets/datasets_SEC/Aggregated SEC Data D Removed First.xlsx"
# df.to_excel(path, index=False, header=True)
# df = pd.read_excel(path)
# display(df.tail(100))



### working example of a function
# def my_func(arg1, arg2):
#     combined = arg1+arg2
#     return combined # this part is unncessary

# answer = my_func(2,4)
# print(answer)



### working example of if statement using PU_count_list in for loop
# for idx, val in enumerate(u_CIK_list[:10]):
#     if PU_count_list[idx] > 0:
#         print(idx, PU_count_list[idx])

    
    
### working example of if-else combo
# for idx, value in enumerate(range(10)):
#     if idx < 5:
#         print('idx less than 5')
#     else:
#         print('idx equal to or greater than 5')




### Exploration of Shared_Sort Function
# def shared_sort():
#     C_count = 0
#     C_row_list = []
#     C_row_success_list = []
#     print('checking row {}'.format(row))
    
#     if column[2] == 'C':
#         C_row_success_list = []
#         C_row_list.append(row)
#         C_row_success_list.append(row)
#         C_count +=1
#         print ("appended row {} to success list".format(row))
#         xdf.iloc[row, -2] = C_count
#         print('Form_C_count: {} added to row {}'.format(C_count, row))
            
#     if column[2] == 'C-W':
#         C_row_success_list.remove(C_row_list[-1])
#         print('removed row {} from success list, new success list: {}'.format(C_row_list[-1], C_row_success_list))
        
#     return C_row_list, C_row_success_list




### Exploration of sorting loop using Shared_Sort Function
# xdf = df.copy()
# ndf = pd.DataFrame().reindex_like(xdf)
# ndf = ndf.dropna()
# display(xdf.head(5))
    
    
# for idx, val in enumerate(u_CIK_list[:2]):
#     if PU_count_list[idx] > 0:        
#         print('CIK {} USING PU METHOD BECAUSE PU_count = {}'.format(val, PU_count_list[idx]))
#         for row, column in xdf[xdf['CIK'] == val].iterrows():
#             C_row_list, C_row_success_list = shared_sort()
#             display(C_row_success_list)
        
#             if column[4] is not np.nan:
#                 xdf.iloc[C_row_success_list, -1] = 1
#                 print('funding success added to row {}'.format(C_row_success_list))
        
#     else:
#         print('CIK {} USING C-AR METHOD BECAUSE PU_count = {}'.format(val, PU_count_list[idx]))
#         C_row_list, C_row_success_list = shared_sort()
       
#         if column[2] == 'C-AR':
#             xdf.iloc[C_row_success_list, -1] = 1
#             print('funding success added to row {}'.format(C_row_success_list))

#     display(xdf[xdf['CIK'] == val])
#     ndf = ndf.append(xdf.iloc[C_row_list,:])
#     C_count = 0
#     C_row_list = []
#     C_row_success_list = []
#     print ('exiting specific CIK and entering new one...')

# print('loop complete')
# ndf = ndf.astype({'CIK': int, 'Funding_Success': int, 'Form_C_Number': int})
# display('Crafted Dataset below')
# display(ndf.head(100))

#### 1. General Imports & Pandas Settings

In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
import datetime
import os


# pd.set_option("max_colwidth", 250)
pd.set_option("max_colwidth", 150)
pd.set_option("display.max_columns", 60)
pd.set_option('max_rows', None)

#### 2. Loading master dataframe from saved Excel

In [2]:
path = "/root/Funders-USA-ML-Backend/datasets/datasets_SEC/Aggregated SEC Data.xlsx"
df = pd.read_excel(path)


df = df.sort_values('FILING_DATE')
df.reset_index(drop=True, inplace=True)


df['FILING_DATE'] = pd.to_datetime(df['FILING_DATE'], format='%Y%m%d')
display(df.shape)
display(df.tail(2))

(9316, 58)

ACCESSION_NUMBER SUBMISSION_TYPE FILING_DATE      CIK FILE_NUMBER  \
9314  0001665160-20-001624             C-U  2020-12-31  1776306   020-25436   
9315  0001666102-20-000034               C  2020-12-31  1817960   020-27423   

      PERIOD    ACCESSION_NUMBER.1  ISAMENDMENT  \
9314     NaN  0001665160-20-001624            0   
9315     NaN  0001666102-20-000034            0   

                                                                                            PROGRESSUPDATE  \
9314  At the close of their offering, the issuer closed on $80,582.34 and 6,928 in shares of common stock.   
9315                                                                                                   NaN   

     NATUREOFAMENDMENT                           NAMEOFISSUER  \
9314               NaN                   Vegan Fine Foods LLC   
9315               NaN  SKYWIDE REAL ESTATE INTERNATIONAL LLC   

                LEGALSTATUSFORM LEGALSTATUSOTHERDESC JURISDICTIONORGANIZATION  \
9314  Limited Liability Company                  NaN                       FL   
9315  Limited Liability Company                  NaN                       PA   

     DATEINCORPORATION            STREET1 STREET2             CITY  \
9314        2017-07-14  330 SW 2ND STREET     NaN  FORT LAUDERDALE   
9315        2020-06-24    1939 MIFFLIN ST     NaN     PHILADELPHIA   

     STATEORCOUNTRY ZIPCODE                          ISSUERWEBSITE  \
9314             FL   33312          http://www.veganfinefoods.com   
9315             PA   19145  https://www.skywideinternational.com/   

              COMPANYNAME  COMMISSIONCIK COMMISSIONFILENUMBER  CRDNUMBER  \
9314  Wefunder Portal LLC      1670254.0            007-00033   283503.0   
9315            Mr. Crowd      1666102.0            007-00042   284278.0   

        ACCESSION_NUMBER.2  \
9314  0001665160-20-001624   
9315  0001666102-20-000034   

                                                                                                                                         COMPENSATIONAMOUNT  \
9314  7.5% of the offering amount upon a successful fundraise, and be entitled to reimbursement for out-of-pocket third party expenses it pays or incurs...   
9315  1) The issuer shall pay the intermediary a fee equal to 5% of the gross amount raised. 2) The issuer shall pay or reimburse all reasonable out-of-...   

     FINANCIALINTEREST SECURITYOFFEREDTYPE SECURITYOFFEREDOTHERDESC  \
9314                No               Other         Convertible Note   
9315               No.               Other         Membership Units   

      NOOFSECURITYOFFERED  PRICE  \
9314              50000.0    1.0   
9315                300.0  100.0   

                                                                                                                                   PRICEDETERMINATIONMETHOD  \
9314  Pro-rated portion of the total principal value of $50,000; interests will be sold in increments of $1; each investment is convertible to one unit ...   
9315                                                                                                                                                    NaN   

      OFFERINGAMOUNT OVERSUBSCRIPTIONACCEPTED  OVERSUBSCRIPTIONALLOCATIONTYPE  \
9314         50000.0                        Y                           Other   
9315         30000.0                        Y  First-come, first-served basis   

             DESCOVERSUBSCRIPTION  MAXIMUMOFFERINGAMOUNT DEADLINEDATE  \
9314  As determined by the issuer              1070000.0   2019-09-27   
9315                          NaN               900000.0   2021-12-31   

      CURRENTEMPLOYEES  TOTALASSETMOSTRECENTFISCALYEAR  \
9314               6.0                        313342.0   
9315               2.0                         25000.0   

      TOTALASSETPRIORFISCALYEAR  CASHEQUIMOSTRECENTFISCALYEAR  \
9314                   144762.0                       62972.0   
9315                        0.0                       250

#### 3. Shrinking dataframe to MVP columns

In [4]:
list_of_cols_MVP = ['CIK', 'FILING_DATE', 'SUBMISSION_TYPE', 'OFFERINGAMOUNT', 'PROGRESSUPDATE', 'NATUREOFAMENDMENT', 'MAXIMUMOFFERINGAMOUNT', 'CURRENTEMPLOYEES', 'TOTALASSETMOSTRECENTFISCALYEAR',
                       'TOTALASSETPRIORFISCALYEAR', 'CASHEQUIMOSTRECENTFISCALYEAR', 'CASHEQUIPRIORFISCALYEAR',
                       'ACTRECEIVEDRECENTFISCALYEAR', 'ACTRECEIVEDPRIORFISCALYEAR', 'SHORTTERMDEBTMRECENTFISCALYEAR',
                       'SHORTTERMDEBTPRIORFISCALYEAR', 'LONGTERMDEBTRECENTFISCALYEAR', 'LONGTERMDEBTPRIORFISCALYEAR',
                       'NETINCOMEMOSTRECENTFISCALYEAR', 'NETINCOMEPRIORFISCALYEAR', 'Form_C_Number', 'Funding_Success']
df = df[list_of_cols_MVP]


display(df.shape)
display(df.head(4))
display(df.tail(1))

(9316, 22)

CIK FILING_DATE SUBMISSION_TYPE  OFFERINGAMOUNT PROGRESSUPDATE  \
0  1674180  2016-05-16               C        275000.0            NaN   
1  1674082  2016-05-16               C        400000.0            NaN   
2  1674632  2016-05-16               C         50000.0            NaN   
3  1674207  2016-05-16               C        250000.0            NaN   

  NATUREOFAMENDMENT  MAXIMUMOFFERINGAMOUNT  CURRENTEMPLOYEES  \
0               NaN               500000.0               3.0   
1               NaN              1000000.0               3.0   
2               NaN              1000000.0              10.0   
3               NaN              1000000.0               3.0   

   TOTALASSETMOSTRECENTFISCALYEAR  TOTALASSETPRIORFISCALYEAR  \
0                             0.0                        0.0   
1                        230909.0                   441532.0   
2                       5015954.0                        0.0   
3                         10170.0                     7321.0   

   CASHEQUIMOSTRECENTFISCALYEAR  CASHEQUIPRIORFISCALYEAR  \
0                           0.0                      0.0   
1                      193376.0                 432140.0   
2                        1250.0                      0.0   
3                       10170.0                   7321.0   

   ACTRECEIVEDRECENTFISCALYEAR  ACTRECEIVEDPRIORFISCALYEAR  \
0                          0.0                         0.0   
1                      -1097.0                         0.0   
2                          0.0                         0.0   
3                          0.0                         0.0   

   SHORTTERMDEBTMRECENTFISCALYEAR  SHORTTERMDEBTPRIORFISCALYEAR  \
0                             0.0                           0.0   
1                        589120.0                      569251.0   
2                        267070.0                           0.0   
3                             0.0                           0.0   

   LONGTERMDEBTRECENTFISCALYEAR  LONGTERMDEBTPRIORFISCALYEAR  \
0                           0.0                          0.0   
1                      230909.0                     441532.0   
2                           0.0                          0.0   
3                           0.0                          0.0   

   NETINCOMEMOSTRECENTFISCALYEAR  NETINCOMEPRIORFISCALYEAR  Form_C_Number  \
0                            0.0                       0.0              0   
1                      -298573.0                 -361237.0              0   
2                      -230547.0                       0.0              0   
3                       -36772.0                   -2250.0              0   

   Funding_Success  
0                0  
1                0  
2                0  
3                0

CIK FILING_DATE SUBMISSION_TYPE  OFFERINGAMOUNT PROGRESSUPDATE  \
9315  1817960  2020-12-31               C         30000.0            NaN   

     NATUREOFAMENDMENT  MAXIMUMOFFERINGAMOUNT  CURRENTEMPLOYEES  \
9315               NaN               900000.0               2.0   

      TOTALASSETMOSTRECENTFISCALYEAR  TOTALASSETPRIORFISCALYEAR  \
9315                         25000.0                        0.0   

      CASHEQUIMOSTRECENTFISCALYEAR  CASHEQUIPRIORFISCALYEAR  \
9315                       25000.0                      0.0   

      ACTRECEIVEDRECENTFISCALYEAR  ACTRECEIVEDPRIORFISCALYEAR  \
9315                          0.0                         0.0   

      SHORTTERMDEBTMRECENTFISCALYEAR  SHORTTERMDEBTPRIORFISCALYEAR  \
9315                             0.0                           0.0   

      LONGTERMDEBTRECENTFISCALYEAR  LONGTERMDEBTPRIORFISCALYEAR  \
9315                           0.0                          0.0   

      NETINCOMEMOSTRECENTFISCALYEAR  NETINCOMEPRIORFISCALYEAR  Form_C_Number  \
9315                            0.0                       0.0              0   

      Funding_Success  
9315                0

#### 4. Add target column to master df

In [3]:
df_length = len(df)
display(df_length)


df = df.assign(Form_C_Number = [0]*df_length)
df = df.assign(Funding_Success = [0]*df_length)
display(df.tail(2))

9316

ACCESSION_NUMBER SUBMISSION_TYPE FILING_DATE      CIK FILE_NUMBER  \
9314  0001665160-20-001624             C-U  2020-12-31  1776306   020-25436   
9315  0001666102-20-000034               C  2020-12-31  1817960   020-27423   

      PERIOD    ACCESSION_NUMBER.1  ISAMENDMENT  \
9314     NaN  0001665160-20-001624            0   
9315     NaN  0001666102-20-000034            0   

                                                                                            PROGRESSUPDATE  \
9314  At the close of their offering, the issuer closed on $80,582.34 and 6,928 in shares of common stock.   
9315                                                                                                   NaN   

     NATUREOFAMENDMENT                           NAMEOFISSUER  \
9314               NaN                   Vegan Fine Foods LLC   
9315               NaN  SKYWIDE REAL ESTATE INTERNATIONAL LLC   

                LEGALSTATUSFORM LEGALSTATUSOTHERDESC JURISDICTIONORGANIZATION  \
9314  Limited Liability Company                  NaN                       FL   
9315  Limited Liability Company                  NaN                       PA   

     DATEINCORPORATION            STREET1 STREET2             CITY  \
9314        2017-07-14  330 SW 2ND STREET     NaN  FORT LAUDERDALE   
9315        2020-06-24    1939 MIFFLIN ST     NaN     PHILADELPHIA   

     STATEORCOUNTRY ZIPCODE                          ISSUERWEBSITE  \
9314             FL   33312          http://www.veganfinefoods.com   
9315             PA   19145  https://www.skywideinternational.com/   

              COMPANYNAME  COMMISSIONCIK COMMISSIONFILENUMBER  CRDNUMBER  \
9314  Wefunder Portal LLC      1670254.0            007-00033   283503.0   
9315            Mr. Crowd      1666102.0            007-00042   284278.0   

        ACCESSION_NUMBER.2  \
9314  0001665160-20-001624   
9315  0001666102-20-000034   

                                                                                                                                         COMPENSATIONAMOUNT  \
9314  7.5% of the offering amount upon a successful fundraise, and be entitled to reimbursement for out-of-pocket third party expenses it pays or incurs...   
9315  1) The issuer shall pay the intermediary a fee equal to 5% of the gross amount raised. 2) The issuer shall pay or reimburse all reasonable out-of-...   

     FINANCIALINTEREST SECURITYOFFEREDTYPE SECURITYOFFEREDOTHERDESC  \
9314                No               Other         Convertible Note   
9315               No.               Other         Membership Units   

      NOOFSECURITYOFFERED  PRICE  \
9314              50000.0    1.0   
9315                300.0  100.0   

                                                                                                                                   PRICEDETERMINATIONMETHOD  \
9314  Pro-rated portion of the total principal value of $50,000; interests will be sold in increments of $1; each investment is convertible to one unit ...   
9315                                                                                                                                                    NaN   

      OFFERINGAMOUNT OVERSUBSCRIPTIONACCEPTED  OVERSUBSCRIPTIONALLOCATIONTYPE  \
9314         50000.0                        Y                           Other   
9315         30000.0                        Y  First-come, first-served basis   

             DESCOVERSUBSCRIPTION  MAXIMUMOFFERINGAMOUNT DEADLINEDATE  \
9314  As determined by the issuer              1070000.0   2019-09-27   
9315                          NaN               900000.0   2021-12-31   

      CURRENTEMPLOYEES  TOTALASSETMOSTRECENTFISCALYEAR  \
9314               6.0                        313342.0   
9315               2.0                         25000.0   

      TOTALASSETPRIORFISCALYEAR  CASHEQUIMOSTRECENTFISCALYEAR  \
9314                   144762.0                       62972.0   
9315                        0.0                       250

#### 5. Setting up custom loops to extract the needed information

##### Loop for creating list of Unique CIKs

In [5]:
u_CIK_list = df['CIK'].unique()
display(u_CIK_list[0:5])

array([1674180, 1674082, 1674632, 1674207, 1674163])

##### Loop for creating list containing the Progress update counts for each CIK

In [6]:
xdf = df.copy()
PU_count = 0
PU_count_list = []


for i in u_CIK_list:
    
    for row, column in xdf[xdf['CIK'] == i].iterrows():
        if column[4] is not np.nan:
            PU_count += 1

    PU_count_list.append(PU_count)
    PU_count = 0
    
    
print(len(PU_count_list))

2931


##### Reversing the PU_count_list for testing

In [74]:
# display(PU_count_list[:10])
# PU_count_list.reverse()
# display(PU_count_list[:10])

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

[1, 1, 0, 1, 5, 0, 0, 3, 1, 0]

### 6. Loop for assignment of target feature (Funding Success)

In [7]:
xdf = df.copy()
ndf = pd.DataFrame().reindex_like(xdf)
ndf = ndf.dropna()
display(xdf.head(2))


C_row_list = []
C_row_success_list = []
C_count = 0
C_cutoff_date = datetime.datetime.strptime('2019-07-01 00:00:00', '%Y-%m-%d %H:%M:%S')


    
for idx, val in enumerate(u_CIK_list):
    if PU_count_list[idx] > 0:        
        
        
        for row, column in xdf[xdf['CIK'] == val].iterrows():
            if column[2] == 'C' and column[1] < C_cutoff_date:
                C_row_success_list = []
                C_row_list.append(row)
                C_row_success_list.append(row)
                C_count +=1
                xdf.iloc[row, -2] = C_count
            
            
            if column[2] == 'C-W' and len(C_row_success_list) > 0:
                C_row_success_list.remove(C_row_list[-1])

                
            if column[4] is not np.nan:
                xdf.iloc[C_row_success_list, -1] = 1        
    
    
    
        for row, column in xdf[xdf['CIK'] == val].iterrows():

            
            if column[2] == 'C' and column[1] < C_cutoff_date:
                C_row_success_list = []
                C_row_list.append(row)
                C_row_success_list.append(row)
                C_count +=1
                xdf.iloc[row, -2] = C_count
            
            
            if column[2] == 'C-W' and len(C_row_success_list) > 0:
                C_row_success_list.remove(C_row_list[-1])
                
                
            if column[2] == 'C-AR':
                xdf.iloc[C_row_success_list, -1] = 1

                
    ndf = ndf.append(xdf.iloc[C_row_list,:])
    C_row_list = []
    C_row_success_list = []
    C_count = 0

    
print('loop complete')
ndf = ndf.astype({'CIK': int, 'Funding_Success': int, 'Form_C_Number': int})
display('Crafted Dataset below')
display(ndf.head(10))

CIK FILING_DATE SUBMISSION_TYPE  OFFERINGAMOUNT PROGRESSUPDATE  \
0  1674180  2016-05-16               C        275000.0            NaN   
1  1674082  2016-05-16               C        400000.0            NaN   

  NATUREOFAMENDMENT  MAXIMUMOFFERINGAMOUNT  CURRENTEMPLOYEES  \
0               NaN               500000.0               3.0   
1               NaN              1000000.0               3.0   

   TOTALASSETMOSTRECENTFISCALYEAR  TOTALASSETPRIORFISCALYEAR  \
0                             0.0                        0.0   
1                        230909.0                   441532.0   

   CASHEQUIMOSTRECENTFISCALYEAR  CASHEQUIPRIORFISCALYEAR  \
0                           0.0                      0.0   
1                      193376.0                 432140.0   

   ACTRECEIVEDRECENTFISCALYEAR  ACTRECEIVEDPRIORFISCALYEAR  \
0                          0.0                         0.0   
1                      -1097.0                         0.0   

   SHORTTERMDEBTMRECENTFISCALYEAR  SHORTTERMDEBTPRIORFISCALYEAR  \
0                             0.0                           0.0   
1                        589120.0                      569251.0   

   LONGTERMDEBTRECENTFISCALYEAR  LONGTERMDEBTPRIORFISCALYEAR  \
0                           0.0                          0.0   
1                      230909.0                     441532.0   

   NETINCOMEMOSTRECENTFISCALYEAR  NETINCOMEPRIORFISCALYEAR  Form_C_Number  \
0                            0.0                       0.0              0   
1                      -298573.0                 -361237.0              0   

   Funding_Success  
0                0  
1                0

loop complete


'Crafted Dataset below'

CIK FILING_DATE SUBMISSION_TYPE  OFFERINGAMOUNT PROGRESSUPDATE  \
0     1674180  2016-05-16               C        275000.0            NaN   
1     1674082  2016-05-16               C        400000.0            NaN   
2     1674632  2016-05-16               C         50000.0            NaN   
3     1674207  2016-05-16               C        250000.0            NaN   
4     1674163  2016-05-16               C        500000.0            NaN   
1060  1674163  2017-07-24               C          1000.0            NaN   
2624  1674163  2018-05-11               C         99840.0            NaN   
10    1674521  2016-05-16               C        100000.0            NaN   
6     1674376  2016-05-16               C        100000.0            NaN   
7     1674439  2016-05-16               C         80000.0            NaN   

     NATUREOFAMENDMENT  MAXIMUMOFFERINGAMOUNT  CURRENTEMPLOYEES  \
0                  NaN               500000.0               3.0   
1                  NaN              1000000.0               3.0   
2                  NaN              1000000.0              10.0   
3                  NaN              1000000.0               3.0   
4                  NaN              1000000.0               2.0   
1060               NaN                70000.0               5.0   
2624               NaN               901888.0               5.0   
10                 NaN              1000000.0               0.0   
6                  NaN              1000000.0               0.0   
7                  NaN               100000.0               2.0   

      TOTALASSETMOSTRECENTFISCALYEAR  TOTALASSETPRIORFISCALYEAR  \
0                                0.0                        0.0   
1                           230909.0                   441532.0   
2                          5015954.0                        0.0   
3                            10170.0                     7321.0   
4                                0.0                        0.0   
1060                        694289.0                        0.0   
2624                       1153234.0                   694289.0   
10                               0.0                        0.0   
6                            54008.0                    54129.0   
7                            35873.0                        0.0   

      CASHEQUIMOSTRECENTFISCALYEAR  CASHEQUIPRIORFISCALYEAR  \
0                              0.0                      0.0   
1                         193376.0                 432140.0   
2                           1250.0                      0.0   
3                          10170.0                   7321.0   
4                              0.0                      0.0   
1060                      638309.0                      0.0   
2624                       58131.0                   1000.0   
10                             0.0                      0.0   
6                           2156.0                   4414.0   
7                          27303.0                      0.0   

      ACTRECEIVEDRECENTFISCALYEAR  ACTRECEIVEDPRIORFISCALYEAR  \
0                             0.0                         0.0   
1                         -1097.0                         0.0   
2                             0.0                         0.0   
3                             0.0                         0.0   
4                             0.0                         0.0   
1060                       1000.0                         0.0   
2624                          0.0                         0.0   
10                            0.0                         0.0   
6                          5180.0                         0.0   
7                             0.0                         0.0   

      SHORTTERMDEBTMRECENTFISCALYEAR  SHORTTERMDEBTPRIORFISCALYEAR  \
0                                0.0                           0.0   
1                           589120.0                      569251.0   
2                           267070.0                           0.0   
3      

### Final custom preprocessing before going into Data Wrangler

In [36]:
display(ndf.tail(7))
nndf = ndf.copy()
display(nndf.tail(7))

# removing entries with offering amounts less than 9k
nndf = nndf[nndf['OFFERINGAMOUNT'] > 9000]
display(len(nndf))

#droping NA values
nndf = nndf.dropna(how='any', subset =["OFFERINGAMOUNT"])
display(len(nndf))

#dropping NA values
nndf = nndf.dropna(how='any', subset =['MAXIMUMOFFERINGAMOUNT'])
display(len(nndf))
display(len(nndf[nndf['CURRENTEMPLOYEES'] > 40]))
display(nndf.tail(7))

# calculating class balance %
pos_cases = nndf[nndf['Funding_Success'] == 1]
total = len(nndf)
success = len(pos_cases)
no_funding = total - success

# printing class balance %
print('total amount of C forms within date cutoff: {}, successes {}, no funding {}'.format(total, success, no_funding))
print(success / total *100)

display(nndf.describe())

CIK FILING_DATE SUBMISSION_TYPE  OFFERINGAMOUNT PROGRESSUPDATE  \
5118  1780597  2019-06-26               C             NaN            NaN   
5120  1779569  2019-06-27               C       100000.00            NaN   
5121  1778755  2019-06-27               C        25000.00            NaN   
5122  1780976  2019-06-27               C        20000.00            NaN   
5124  1778104  2019-06-28               C        10000.08            NaN   
5127  1781049  2019-06-28               C        50000.00            NaN   
5129  1779032  2019-06-28               C        25000.00            NaN   

     NATUREOFAMENDMENT  MAXIMUMOFFERINGAMOUNT  CURRENTEMPLOYEES  \
5118               NaN                    NaN               NaN   
5120               NaN               200000.0              17.0   
5121               NaN              1070000.0              12.0   
5122               NaN                40000.0               1.0   
5124               NaN               250000.2               2.0   
5127               NaN               107000.0              13.0   
5129               NaN              1070000.0               1.0   

      TOTALASSETMOSTRECENTFISCALYEAR  TOTALASSETPRIORFISCALYEAR  \
5118                             NaN                        NaN   
5120                       3571162.0                        0.0   
5121                       2205822.0                  1006095.0   
5122                             0.0                        0.0   
5124                          1202.0                    12684.0   
5127                         62912.0                    44799.0   
5129                         81172.0                    39720.0   

      CASHEQUIMOSTRECENTFISCALYEAR  CASHEQUIPRIORFISCALYEAR  \
5118                           NaN                      NaN   
5120                       48573.0                      0.0   
5121                     1990076.0                 844610.0   
5122                           0.0                      0.0   
5124                        1202.0                  12684.0   
5127                       43812.0                  32326.0   
5129                       17199.0                   4243.0   

      ACTRECEIVEDRECENTFISCALYEAR  ACTRECEIVEDPRIORFISCALYEAR  \
5118                          NaN                         NaN   
5120                          0.0                         0.0   
5121                      13061.0                     44366.0   
5122                          0.0                         0.0   
5124                          0.0                         0.0   
5127                          0.0                         0.0   
5129                      12887.0                     12071.0   

      SHORTTERMDEBTMRECENTFISCALYEAR  SHORTTERMDEBTPRIORFISCALYEAR  \
5118                             NaN                           NaN   
5120                          5100.0                           0.0   
5121                       9037325.0                      327548.0   
5122                             0.0                           0.0   
5124                             0.0                           0.0   
5127                         70877.0                       11406.0   
5129                         73313.0                       34488.0   

      LONGTERMDEBTRECENTFISCALYEAR  LONGTERMDEBTPRIORFISCALYEAR  \
5118                           NaN                          NaN   
5120                     3600000.0                          0.0   
5121                     1003574.0                    5009988.0   
5122                           0.0                          0.0   
5124                           0.0                          0.0   
5127                           0.0                          0.0   
5129                           0.0                          0.0   

      NETINCOMEMOSTRECENTFISCALYEAR  NETINCOMEPRIORFISCALYEAR  Form_C_Number  \
5118                            NaN                       NaN              1   
5120                       -33938.0        

CIK FILING_DATE SUBMISSION_TYPE  OFFERINGAMOUNT PROGRESSUPDATE  \
5118  1780597  2019-06-26               C             NaN            NaN   
5120  1779569  2019-06-27               C       100000.00            NaN   
5121  1778755  2019-06-27               C        25000.00            NaN   
5122  1780976  2019-06-27               C        20000.00            NaN   
5124  1778104  2019-06-28               C        10000.08            NaN   
5127  1781049  2019-06-28               C        50000.00            NaN   
5129  1779032  2019-06-28               C        25000.00            NaN   

     NATUREOFAMENDMENT  MAXIMUMOFFERINGAMOUNT  CURRENTEMPLOYEES  \
5118               NaN                    NaN               NaN   
5120               NaN               200000.0              17.0   
5121               NaN              1070000.0              12.0   
5122               NaN                40000.0               1.0   
5124               NaN               250000.2               2.0   
5127               NaN               107000.0              13.0   
5129               NaN              1070000.0               1.0   

      TOTALASSETMOSTRECENTFISCALYEAR  TOTALASSETPRIORFISCALYEAR  \
5118                             NaN                        NaN   
5120                       3571162.0                        0.0   
5121                       2205822.0                  1006095.0   
5122                             0.0                        0.0   
5124                          1202.0                    12684.0   
5127                         62912.0                    44799.0   
5129                         81172.0                    39720.0   

      CASHEQUIMOSTRECENTFISCALYEAR  CASHEQUIPRIORFISCALYEAR  \
5118                           NaN                      NaN   
5120                       48573.0                      0.0   
5121                     1990076.0                 844610.0   
5122                           0.0                      0.0   
5124                        1202.0                  12684.0   
5127                       43812.0                  32326.0   
5129                       17199.0                   4243.0   

      ACTRECEIVEDRECENTFISCALYEAR  ACTRECEIVEDPRIORFISCALYEAR  \
5118                          NaN                         NaN   
5120                          0.0                         0.0   
5121                      13061.0                     44366.0   
5122                          0.0                         0.0   
5124                          0.0                         0.0   
5127                          0.0                         0.0   
5129                      12887.0                     12071.0   

      SHORTTERMDEBTMRECENTFISCALYEAR  SHORTTERMDEBTPRIORFISCALYEAR  \
5118                             NaN                           NaN   
5120                          5100.0                           0.0   
5121                       9037325.0                      327548.0   
5122                             0.0                           0.0   
5124                             0.0                           0.0   
5127                         70877.0                       11406.0   
5129                         73313.0                       34488.0   

      LONGTERMDEBTRECENTFISCALYEAR  LONGTERMDEBTPRIORFISCALYEAR  \
5118                           NaN                          NaN   
5120                     3600000.0                          0.0   
5121                     1003574.0                    5009988.0   
5122                           0.0                          0.0   
5124                           0.0                          0.0   
5127                           0.0                          0.0   
5129                           0.0                          0.0   

      NETINCOMEMOSTRECENTFISCALYEAR  NETINCOMEPRIORFISCALYEAR  Form_C_Number  \
5118                            NaN                       NaN              1   
5120                       -33938.0        

1753

1739

1721

19

CIK FILING_DATE SUBMISSION_TYPE  OFFERINGAMOUNT PROGRESSUPDATE  \
5109  1779469  2019-06-25               C       800000.00            NaN   
5120  1779569  2019-06-27               C       100000.00            NaN   
5121  1778755  2019-06-27               C        25000.00            NaN   
5122  1780976  2019-06-27               C        20000.00            NaN   
5124  1778104  2019-06-28               C        10000.08            NaN   
5127  1781049  2019-06-28               C        50000.00            NaN   
5129  1779032  2019-06-28               C        25000.00            NaN   

     NATUREOFAMENDMENT  MAXIMUMOFFERINGAMOUNT  CURRENTEMPLOYEES  \
5109               NaN              1070000.0               0.0   
5120               NaN               200000.0              17.0   
5121               NaN              1070000.0              12.0   
5122               NaN                40000.0               1.0   
5124               NaN               250000.2               2.0   
5127               NaN               107000.0              13.0   
5129               NaN              1070000.0               1.0   

      TOTALASSETMOSTRECENTFISCALYEAR  TOTALASSETPRIORFISCALYEAR  \
5109                             0.0                        0.0   
5120                       3571162.0                        0.0   
5121                       2205822.0                  1006095.0   
5122                             0.0                        0.0   
5124                          1202.0                    12684.0   
5127                         62912.0                    44799.0   
5129                         81172.0                    39720.0   

      CASHEQUIMOSTRECENTFISCALYEAR  CASHEQUIPRIORFISCALYEAR  \
5109                           0.0                      0.0   
5120                       48573.0                      0.0   
5121                     1990076.0                 844610.0   
5122                           0.0                      0.0   
5124                        1202.0                  12684.0   
5127                       43812.0                  32326.0   
5129                       17199.0                   4243.0   

      ACTRECEIVEDRECENTFISCALYEAR  ACTRECEIVEDPRIORFISCALYEAR  \
5109                          0.0                         0.0   
5120                          0.0                         0.0   
5121                      13061.0                     44366.0   
5122                          0.0                         0.0   
5124                          0.0                         0.0   
5127                          0.0                         0.0   
5129                      12887.0                     12071.0   

      SHORTTERMDEBTMRECENTFISCALYEAR  SHORTTERMDEBTPRIORFISCALYEAR  \
5109                           800.0                           0.0   
5120                          5100.0                           0.0   
5121                       9037325.0                      327548.0   
5122                             0.0                           0.0   
5124                             0.0                           0.0   
5127                         70877.0                       11406.0   
5129                         73313.0                       34488.0   

      LONGTERMDEBTRECENTFISCALYEAR  LONGTERMDEBTPRIORFISCALYEAR  \
5109                           0.0                          0.0   
5120                     3600000.0                          0.0   
5121                     1003574.0                    5009988.0   
5122                           0.0                          0.0   
5124                           0.0                          0.0   
5127                           0.0                          0.0   
5129                           0.0                          0.0   

      NETINCOMEMOSTRECENTFISCALYEAR  NETINCOMEPRIORFISCALYEAR  Form_C_Number  \
5109                            0.0                       0.0              1   
5120                       -33938.0        

total amount of C forms within date cutoff: 1721, successes 902, no funding 819
52.411388727484024


CIK  OFFERINGAMOUNT  MAXIMUMOFFERINGAMOUNT  CURRENTEMPLOYEES  \
count  1.721000e+03    1.721000e+03           1.721000e+03       1721.000000   
mean   1.714285e+06    6.614720e+04           5.845282e+05          5.490413   
std    6.422606e+04    1.023510e+05           4.272982e+05         10.783495   
min    3.529560e+05    9.500000e+03           1.000000e+02          0.000000   
25%    1.695260e+06    1.000000e+04           1.070000e+05          1.000000   
50%    1.722315e+06    2.500000e+04           5.000000e+05          3.000000   
75%    1.748332e+06    7.500000e+04           1.000000e+06          6.000000   
max    1.781049e+06    1.070000e+06           1.070000e+06        225.000000   

       TOTALASSETMOSTRECENTFISCALYEAR  TOTALASSETPRIORFISCALYEAR  \
count                    1.721000e+03               1.721000e+03   
mean                     5.176145e+05               3.706793e+05   
std                      7.409033e+06               6.159633e+06   
min                     -6.170000e+02              -2.163700e+04   
25%                      0.000000e+00               0.000000e+00   
50%                      2.498000e+04               0.000000e+00   
75%                      2.006744e+05               8.070000e+04   
max                      2.984769e+08               2.478195e+08   

       CASHEQUIMOSTRECENTFISCALYEAR  CASHEQUIPRIORFISCALYEAR  \
count                  1.721000e+03             1.721000e+03   
mean                   8.688630e+04             5.636765e+04   
std                    3.427203e+05             2.914057e+05   
min                   -6.220800e+04            -8.829300e+04   
25%                    0.000000e+00             0.000000e+00   
50%                    3.422000e+03             0.000000e+00   
75%                    3.697500e+04             1.465200e+04   
max                    7.101233e+06             8.783731e+06   

       ACTRECEIVEDRECENTFISCALYEAR  ACTRECEIVEDPRIORFISCALYEAR  \
count                 1.721000e+03                1.721000e+03   
mean                  1.819310e+04                1.481553e+04   
std                   1.373676e+05                1.289064e+05   
min                  -1.114112e+04               -4.667000e+03   
25%                   0.000000e+00                0.000000e+00   
50%                   0.000000e+00                0.000000e+00   
75%                   0.000000e+00                0.000000e+00   
max                   4.765020e+06                4.277674e+06   

       SHORTTERMDEBTMRECENTFISCALYEAR  SHORTTERMDEBTPRIORFISCALYEAR  \
count                    1.721000e+03                  1.721000e+03   
mean                     1.700995e+05                  1.040928e+05   
std                      1.197809e+06                  5.829421e+05   
min                     -2.065250e+05                 -5.901200e+02   
25%                      0.000000e+00                  0.000000e+00   
50%                      0.000000e+00                  0.000000e+00   
75%                      5.886200e+04                  1.224700e+04   
max                      4.466959e+07                  1.844276e+07   

       LONGTERMDEBTRECENTFISCALYEAR  LONGTERMDEBTPRIORFISCALYEAR  \
count                  1.721000e+03                 1.721000e+03   
mean                   2.898485e+05                 2.330814e+05   
std                    4.054356e+06                 4.022068e+06   
min                   -4.448429e+04                 0.000000e+00   
25%                    0.000000e+00                 0.000000e+00   
50%                    0.000000e+00                 0.000000e+00   
75%                    6.330500e+04                 3.352000e+03   
max                    1.655286e+08                 1.648934e+08   

       NETINCOMEMOSTRECENTFISCALYEAR  NETINCOMEPRIORFISCALYEAR  Form_C_Number  \
count                   1.721000e+03              1.721000e+03    1721.000000   
mean                   -1.930009e+05             -1.189076e+05       1.115630   
std           

### Loading crafted df to excel, and then to S3 bucket

In [9]:
import boto3
file_path = "/root/Funders-USA-ML-Backend/datasets/datasets_SEC/MVP_Success_Assigned_CSV.csv"
bucket = 'funders-usa-sec-datasets'
prefix = 'datasets_SEC/'
new_name_of_file = 's3_upload_test_1_SA_CSV'

s3_client = boto3.client("s3")
s3_client.upload_file(file_path, bucket, f"{prefix}/{new_name_of_file}.csv")

S3UploadFailedError: Failed to upload /root/Funders-USA-ML-Backend/datasets/datasets_SEC/MVP_Success_Assigned_CSV.csv to funders-usa-sec-datasets/datasets_SEC//s3_upload_test_1_SA_CSV.csv: An error occurred (AccessDenied) when calling the PutObject operation: Access Denied

<br>

### RL: MVP classification model
#### Entries to keep:
* Latest “C-AR” entry of unique CIK - “1”
* Latest entry of unique CIK - “0”
    * Question on this one, do all companies file a C/W if they do not raise? I don't think so, if they don't, then I will keep what I said above. If they do then I will keep latest “C/W” entry of unique CIK - “0”
            
1. Once I have this list, I can 100% run the MVP in Autopilot.
1. Okay so what's the first step? - Being able to do this check, and delete, for 1 CIK. - heading over to code
1. Awesome, SEC posted 2020Q4, instantly added it into the master df. That worked super awesomely as expected, but, for some reason the same code in SM was putting them out of order, so I used the code from my local path folder. nbd. Done. Great, so now we have 2931 unique companies in here. Great day today.
1. Starting off next session: (1 step at time) - create list of unique CIK values. Done.
1. SoN - funders USA RL, A. I will test out AWS Data Wrangler B. I will test out AWS Clarify, C. I will continue with using the .drop_duplicates function to easily get 2900 unique companies in here.
1. Okay, so DW is a wellbuilt GUI for doing data transforms, pick the transform type, pick the column and pick the settings and wallah. So Im wondering if I can do all the transforms here, or if there are a couple that may be easier in the notebook...
    * drop duplicates - so before completeing this step, I want to basically make sure we're capturing the changes in the data that are getting dropped, however, after this step, I have a list of pretty uniform changes to make across the board, I think thats where DW would come in very handy.
    * Numerical encoding - this is definitely a very easy thing in DW
    * Scaling values - Im pretty sure that Autopilot has this built in, so I dont even think its necessary here?
    * Setting File_date with hyphens as a date - this one looks like it can be done here too, just with a few iterations of "find and replace"
    * also going to want to shuffle this data, and find the class imbalance
1. Okay cool, so I think it would make sense to do this initial part of removing the duplicates and storing the values in the notebook, I could do it in this interface as well, but it would be easier to see the code, and cost less in the notebook. After that, I think DW would be great to handle all the smaller, uniform transformations.
1. Wow, okay what a day, so the drop_duplicates function is now working perfectly as long as I save as an excel and reopen. So. Moving forward, what I will do is: I will create a loop which creates a new column to record each of the important changes, such as offering amount, for each company - the first thing I will do is write out this list of potential new columns, after understanding this list, I will add them to the final entry of each company, once this is complete, I simply drop_duplicates, keeping the last entry, and all the correct information for each company will be kept. Thats literally it. all unique companies with all the relevant changes overtime. After this point I will simply be able to go into DW, and do the final, simple transformations. So - starting off - one line at a time - look at this dataset and create a list of potential new columns to be made. after this initial list, just start with the first line of code to get the first one working, all others will slide right in after this. Awesome. This is really coming together.



# RL for integrating into Sagemaker
##### Cost = less than $1 per day. Done.
##### Path of files. Done.
##### Toggle "collapse cell". Done.


# RUNNING LOG for Dataset cleaning:
##### Okay great, so first step is going to be to keep only the 27 initial columns. Done.
##### Okay so let's just create a column called “Funding_Success”. Done.
##### Next, I wrote a function which loops through each row in the dataset, checking for the appearance of "C-AR", if its there, then Im going to put a "1" into the Funding_Success column, if not Im going to put a 0. Done.
##### Once Ive got that, I can remove all columns except for the 16 numeric ones. Done.


# RUNNING LOG FOR EXCEL MERGING:
##### hmm okay in plain english, what is the order of operations here? last: merge 3 docs into one data frame, next, join all the rows of that dataframe to a master dataframe, repeat this 20 times for all quarters. Done.
######  - OKAY SO HERES WHERE TO START OFF - BASICALLY, THIS LOOP ABOVE IS WORKING VERY WELL FOR WHAT IM ASKING IT TO DO, THE FIRST STEP I WILL COME BACK TO, IS THE READ_TABLE() FUNCTION, THIS FUNCTION IS LOOKING TO BE PASSED A FILE PATH, AND CURRENTLY I AM PASSING IT A STRING LIST, SO ALL I NEED TO DO IS CREATE A STRING LIST THAT CONTAINS THE ENTIRE FILE'S PATH, AND THIS WILL WORK. FOCUS ON ONE STEP AT A TIME.   ----------------------- AFTER THIS, I WILL CONTINUE BY USING THE PROPER SYNTAX FOR THE "MERGE" COMMAND IN PANDAS TO GET THE FIRST QUARTER INTO ONE DATAFRAME
######  - Starting 1/11, okay so earlier I was working on a function which created a list of folder names, and then I learned how to add/remove characters to each item in a list. Thats great and all, but I need it to actually rename the files. now I could spend more time on this file renaming part, or just do it manually and move on. I think its worth atleast looking into.
###### - Okay so Im able to read the tsv files into a dataframe in a loop. Done.
###### - Now I want to be able to merge these dataframes together, so lets take this out of the loop and get this command to work on its own.
###### - NEWEST START OFF - Okay so Im now able to read these tsv files into a dataframe, and successfull concat these dataframes together, awesome, now Im realizing that I only want to do this with 3 of the 6 files. So the next step is rewriting the loop, so that it does this with only the 3 I need, where the rows all line up. The loop worked perfectly with all the documents, now I am just running the loop with 3 instead of 6. ONE STEP AT A TIME. ----------------------- AFTER THIS, I will have a dataframe that is the entire quarter, merged together and aligned. PERFECT. After this, I will write the outer loop which does this for all quarters, that should be easy, and once I have all quarters in separate dfs, I will write another outer loop on top of this, which "appends" all these quarters together, or just do it manually if its just too much lol. At that point Ill be done with the data pulling loop, and Ill have the master dataset of 5 years from these 3 documents, all combined.
###### Beautiful, so Ive got all 3 documents in the first quarter lined up perfectly!
###### Okay, so right now I am specifying a folder name, and then running a loop to add the file path onto the contents of the folder...
###### What I need to do, is get a list of all folder PATHS, the file PATHS loop is already created
###### Okay so I know that if I just change the slashes it works with the os.walk() function. Wow. Okay, but apparently backslashes are an "escape" character in python, so a simple replace command will not work. 
###### Alright then, Im going to call it here, basically, Im making progress on merging these, I have an inner loop that works perfectly to merge all datasets in a quarter, so Im just simply doing whats necessary to create an outer "for i in dirs_paths:" line. The next step involves 
###### Okay, so START HERE - Figuring out how to really use the OS library is going to be key for moving forward. 
###### Great, so I have a working example of the append command, so now just time to see where the loop is looping
###### Done. Complete dataset of 8300 entries. Awesome work.

#### Notes for Later Steps
Going to remove any entries from master_df WHERE filing_date is > 2019-10-01 WHERE not C-AR forms (something like that) so we dont have any businesses which dont have the target var.